pkg/scraper/scraper.go

```diff
@@ -83,110 +77,145 @@ func RegisterScraperMetrics(registrationFunc func(metrics.Registerable) error) e
+	scrapeTimeout    time.Duration
+	metricResolution time.Duration
+	// Tracks all running per-node goroutines - per-node goroutine will be
+	// processing updates received through its corresponding channel.
+	stop map[apitypes.UID]chan struct{}
+	// Track the current state of per-node goroutines.
+	// Currently all update request for a given node coming when another
+	// update of this node is being processed are ignored.
+	isWorking map[apitypes.UID]bool
+
+	storage storage.Storage
 }
 
-var _ Scraper = (*scraper)(nil)
-
 // NodeInfo contains the information needed to identify and connect to a particular node
 // (node name and preferred address).
 type NodeInfo struct {
 	Name           string
 	ConnectAddress string
 }
 
-func (c *scraper) Scrape(baseCtx context.Context) *storage.MetricsBatch {
-	nodes, err := c.nodeLister.List(labels.Everything())
-	if err != nil {
-		// report the error and continue on in case of partial results
-		klog.ErrorS(err, "Failed to list nodes")
+func (m *manageNodeScraper) AddNodeScraper(node *corev1.Node) error {
+	m.nodeLock.Lock()
+	defer m.nodeLock.Unlock()
+	if working, exists := m.isWorking[node.UID]; exists && working {
+		klog.V(1).ErrorS(fmt.Errorf("Scrape in node is already running"), "", "node", klog.KObj(node))
+		return fmt.Errorf("Scrape in node is already running, node:%v", node)
 	}
-	klog.V(1).InfoS("Scraping metrics from nodes", "nodeCount", len(nodes))
-
-	responseChannel := make(chan *storage.MetricsBatch, len(nodes))
-	defer close(responseChannel)
-
-	startTime := myClock.Now()
-
-	// TODO(serathius): re-evaluate this code -- do we really need to stagger fetches like this?
-	delayMs := delayPerSourceMs * len(nodes)
-	if delayMs > maxDelayMs {
-		delayMs = maxDelayMs
-	}
-
-	for _, node := range nodes {
-		go func(node *corev1.Node) {
-			// Prevents network congestion.
-			sleepDuration := time.Duration(rand.Intn(delayMs)) * time.Millisecond
-			time.Sleep(sleepDuration)
-			// make the timeout a bit shorter to account for staggering, so we still preserve
-			// the overall timeout
-			ctx, cancelTimeout := context.WithTimeout(baseCtx, c.scrapeTimeout-sleepDuration)
-			defer cancelTimeout()
-			klog.V(2).InfoS("Scraping node", "node", klog.KObj(node))
-			m, err := c.collectNode(ctx, node)
-			if err != nil {
-				klog.ErrorS(err, "Failed to scrape node", "node", klog.KObj(node))
-			}
-			responseChannel <- m
-		}(node)
+	klog.V(1).InfoS("Start scrape metrics from node", "node", klog.KObj(node))
+	stopCh, exists := m.stop[node.UID]
+	if !exists {
+		stopCh = make(chan struct{})
+		m.stop[node.UID] = stopCh
 	}
-
-	res := &storage.MetricsBatch{
-		Nodes: map[string]storage.MetricsPoint{},
-		Pods:  map[apitypes.NamespacedName]storage.PodMetricsPoint{},
-	}
-
-	for range nodes {
-		srcBatch := <-responseChannel
-		if srcBatch == nil {
-			continue
+	go func() {
+		ticker := time.NewTicker(m.metricResolution)
+		defer ticker.Stop()
+
+		res, _ := m.ScrapeData(node)
```

Comment:
Why return error from `ScrapeData` if it's silenced everywhere?

Reply:
I will update it.